### PR TITLE
Add fragment helper to render them in front

### PR DESCRIPTION
**Pull request by srascar**
## Changelog

``` markdown
### Added
- Added `TwigExtension` to wrap `FragmentHelper` in twig templates
- Added `FragmentHelper` to render Fragments
- Associated Unit Tests and Doc
- Added `FragmentService::getTemplate()`
```
## Subject

This PR comes with everything you need to handle the rendering of article's fragments

### CODE REVIEW

**Comment by greg0ire** [comment]
Please associate your email address with your github account

File: DependencyInjection/CompilerPass/FragmentsCompilerPass.php [hunk 1]
```diff
@@ -26,7 +26,7 @@
     public function process(ContainerBuilder $container)
     {
         $fragmentServiceIds = $container->findTaggedServiceIds('sonata.article.fragment');
-        $fragmentServices = array();
+        $fragmentServices = [];
```
Comment (greg0ire) [inline DependencyInjection/CompilerPass/FragmentsCompilerPass.php]: Should be in a separate commit.

File: DependencyInjection/CompilerPass/FragmentsCompilerPass.php [hunk 2]
```diff
@@ -38,7 +38,12 @@ public function process(ContainerBuilder $container)
 
         if ($container->hasDefinition('sonata.article.admin.fragment')) {
             $fragmentAdminDef = $container->getDefinition('sonata.article.admin.fragment');
-            $fragmentAdminDef->addMethodCall('setFragmentServices', array($fragmentServices));
+            $fragmentAdminDef->addMethodCall('setFragmentServices', [$fragmentServices]);
```
Comment (greg0ire) [inline DependencyInjection/CompilerPass/FragmentsCompilerPass.php]: This too. Use `git add --patch` to commit parts of a file.

File: Helper/FragmentHelper.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    public function render(FragmentInterface $fragment)
+    {
+        $type = $fragment->getType();
+
+        if (!array_key_exists($type, $this->fragmentServices)) {
+            throw new \RuntimeException(sprintf('Cannot render Fragment of type `%s`. Service not found.', $type));
+        }
+
+        $fragmentService = $this->fragmentServices[$type];
+        $templateName = $fragmentService->getTemplate();
```
Comment (greg0ire) [inline Helper/FragmentHelper.php]: One-time variable, no need for that.

File: Helper/FragmentHelper.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    public function render(FragmentInterface $fragment)
+    {
+        $type = $fragment->getType();
+
+        if (!array_key_exists($type, $this->fragmentServices)) {
+            throw new \RuntimeException(sprintf('Cannot render Fragment of type `%s`. Service not found.', $type));
+        }
+
+        $fragmentService = $this->fragmentServices[$type];
```
Comment (greg0ire) [inline Helper/FragmentHelper.php]: One time variable

File: Helper/FragmentHelper.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        $templateName = $fragmentService->getTemplate();
+
+        return $this->templating->render(
+            $templateName,
```
Comment (greg0ire) [inline Helper/FragmentHelper.php]: Simply do `$this->fragmentServices[$type]->getTemplate()`, don't make us look up to see where and how this variable is defined.

File: Resources/config/helper.xml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="sonata.article.helper.fragment" class="Sonata\ArticleBundle\Helper\FragmentHelper">
+            <argument type="service" id="templating"/>
+        </service>
+    </services>
+</container>
```
Comment (greg0ire) [inline Resources/config/helper.xml]: Please configure your IDE to always add a newline, it will make this red pictogram disappear.

Comment (greg0ire) [inline Resources/config/helper.xml]: Same for other files below please.

File: Resources/doc/reference/custom_fragment.rst [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        tags:
+            - { name: sonata.article.fragment, key: acme.dummy.fragment.my_awesome }
+            # Where key is your fragment type unique identifier
+
+
+Define a fragment template
+--------------------------
+
+To render the fragment, simply create a template as you defined it in the service.
+Using the twig helper you will be able to access the following variable inside this template:
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: - add a comma after "helper"
- "variable_s_"

File: Resources/doc/reference/custom_fragment.rst [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+* ``fragment`` : The full fragment object.
+* ``elements`` : The values that was set in fragment settings.
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: was => were

File: Resources/doc/reference/getting_started.rst [hunk 9]
```diff
@@ -99,11 +99,53 @@ Render article fragments
 ------------------------
 
 **This part is currently in development.**
-The goal is to provide a simple twig helper which will render each fragments associated to an article.
-Each fragment has its own template, this twig function is going to need to fetch in ``FragmentService`` the template,
-then render it.
+SonataArticleBundle now comes with a twig helper which allows you to render article's fragments
+if they are enable.
```
Comment (greg0ire) [inline Resources/doc/reference/getting_started.rst]: "enable_d_"

File: Resources/doc/reference/getting_started.rst [hunk 10]
```diff
@@ -99,11 +99,53 @@ Render article fragments
 ------------------------
 
 **This part is currently in development.**
-The goal is to provide a simple twig helper which will render each fragments associated to an article.
-Each fragment has its own template, this twig function is going to need to fetch in ``FragmentService`` the template,
-then render it.
+SonataArticleBundle now comes with a twig helper which allows you to render article's fragments
```
Comment (greg0ire) [inline Resources/doc/reference/getting_started.rst]: "article fragments"

File: Resources/doc/reference/getting_started.rst [hunk 11]
```diff
@@ -99,11 +99,53 @@ Render article fragments
...
+if they are enable.
 
 
 .. code-block:: jinja
 
-    {{ sonata_article_render_fragments(article) }}
+    {{ sonata_article_render_article_fragments(article) }}
+
+
+Or a specific fragment whether it is enable or not.
```
Comment (greg0ire) [inline Resources/doc/reference/getting_started.rst]: "enabled"

File: Resources/doc/reference/getting_started.rst [hunk 12]
```diff
@@ -99,11 +99,53 @@ Render article fragments
...
-Each fragment has its own template, this twig function is going to need to fetch in ``FragmentService`` the template,
-then render it.
+SonataArticleBundle now comes with a twig helper which allows you to render article's fragments
+if they are enable.
 
 
 .. code-block:: jinja
 
-    {{ sonata_article_render_fragments(article) }}
+    {{ sonata_article_render_article_fragments(article) }}
```
Comment (greg0ire) [inline Resources/doc/reference/getting_started.rst]: maybe `sonata_article_render_fragments` would be unique enough? Especially since `article` is passed as argument.

Comment (greg0ire) [inline Resources/doc/reference/getting_started.rst]: Oh, I see why you did this. Nevermind.

File: Resources/doc/reference/getting_started.rst [hunk 13]
```diff
@@ -99,11 +99,53 @@ Render article fragments
...
+
+Or a specific fragment whether it is enable or not.
+
+
+.. code-block:: jinja
+
+    {{ sonata_article_render_fragment(article.fragments[0]) }}
+
+
+This extension is based on the FragmentHelper class so you can access the fragment render directly
```
Comment (greg0ire) [inline Resources/doc/reference/getting_started.rst]: "so you can also render fragments" ?

File: Resources/views/Fragment/fragment_text.html.twig [hunk 14]
```diff
@@ -0,0 +1,3 @@
+<p>
+    {{ elements.text|raw }}
+</p>
```
Comment (greg0ire) [inline Resources/views/Fragment/fragment_text.html.twig]: Those are the files I was talking about

File: Tests/Helper/FragmentHelperTest.php [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        $fragmentService->expects($this->once())->method('getTemplate')->will($this->returnValue('template.html.twig'));
+
+        $this->fragmentHelper->setFragmentServices(array('foo.bar' => $fragmentService));
+        $this->fragmentHelper->render($fragment);
+    }
+
+    /**
+     * @return \PHPUnit_Framework_MockObject_MockObject
+     */
+    protected function getFragmentMock()
```
Comment (greg0ire) [inline Tests/Helper/FragmentHelperTest.php]: private ?

File: Tests/Twig/FragmentExtensionTest.php [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+        // we expect only two calls
+        $this->fragmentHelper->expects($this->at(0))
+            ->method('render')
+            ->willReturnCallback(array($this, 'renderFragment'));
+        $this->fragmentHelper->expects($this->at(1))
+            ->method('render')
+            ->willReturnCallback(array($this, 'renderFragment'));
+
+        $this->assertEquals('<h1>foo0</h1><p>bar0</p><h1>foo2</h1><p>bar2</p>', $this->fragmentExtension->renderArticleFragments($article));
```
Comment (greg0ire) [inline Tests/Twig/FragmentExtensionTest.php]: Please linebreak this.

File: Twig/FragmentExtension.php [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        $this->fragmentHelper = $fragmentHelper;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFunctions()
+    {
+        return array(
+            new \Twig_SimpleFunction('sonata_article_render_fragment', array($this, 'renderFragment'), array('is_safe' => array('html'))),
```
Comment (greg0ire) [inline Twig/FragmentExtension.php]: Please linebreak both lines.

**Comment by srascar** [comment]
@greg0ire I just fixed the travis build.
Can you please rerun the check?
Thanks in advance

**Comment by greg0ire** [comment]
Uuuuh… I shouldn't have to, it triggers automatically as soon as you push. Did you push?

**Comment by srascar** [comment]
yes and i checked the diff to be sure

File: Resources/doc/reference/custom_fragment.rst [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+
+Define a fragment Service
+-------------------------
+
+First you need to create a class that extends ``Sonata\ArticleBundle\FragmentService\AbstractFragmentService``
+
+
+.. code-block:: php
+
+namespace Acme\DummyBundle\FragmentService;
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: Not sure you fixed [this](https://travis-ci.org/sonata-project/SonataArticleBundle/jobs/149016747#L292). Shouldn't the block from 13 to 45 be indented?

**Comment by greg0ire** [comment]
I'll just rerun the jobs that failed.

**Comment by greg0ire** [comment]
> Please associate your email address with your github account

Also, please do that if you want the authorship of these commits.

**Comment by srascar** [comment]
you're right sorry

**Comment by greg0ire** [comment]
`docs` is good, now `lint`

**Comment by greg0ire** [comment]
[Not much to fix](https://travis-ci.org/sonata-project/SonataArticleBundle/jobs/149025475#L239)

File: Helper/FragmentHelper.php [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+        if (!array_key_exists($type, $this->fragmentServices)) {
+            throw new \RuntimeException(sprintf('Cannot render Fragment of type `%s`. Service not found.', $type));
+        }
+
+        return $this->templating->render(
+            $this->fragmentServices[$type]->getTemplate(),
+            array(
+                'fragment' => $fragment,
+                'elements' => $fragment->getSettings(),
```
Comment (qsomazzi) [inline Helper/FragmentHelper.php]: not fan of renaming settings to `elements`, we should keep consistancy and name it `settings`

File: Resources/doc/reference/getting_started.rst [hunk 20]
```diff
@@ -99,11 +99,53 @@ Render article fragments
 ------------------------
 
 **This part is currently in development.**
```
Comment (qsomazzi) [inline Resources/doc/reference/getting_started.rst:101]: you can remove this line now

File: Helper/FragmentHelper.php (hunk 19 above)
Comment (srascar) [inline Helper/FragmentHelper.php]: Good notice. I would actually like to rename `settings` into `elements`.
@greg0ire What do you think?

File: Resources/doc/reference/getting_started.rst (hunk 20 above)
Comment (srascar) [inline Resources/doc/reference/getting_started.rst:101]: Regarding the comment above, this should stay

File: Helper/FragmentHelper.php (hunk 19 above)
Comment (greg0ire) [inline Helper/FragmentHelper.php]: The more meaningful the better. What does `$fragment->getSettings()` contain?

Comment (srascar) [inline Helper/FragmentHelper.php]: It does not contains any config. The wording is a legacy from SonataBlock

Comment (greg0ire) [inline Helper/FragmentHelper.php]: Ok, so, not `settings` if possible. Now that I know what it does not contain, may I know what it does contain?

Comment (srascar) [inline Helper/FragmentHelper.php]: Fragments values. It looks like:
`{title:"foo", body:"bar", img:"foo.png"}`

Comment (greg0ire) [inline Helper/FragmentHelper.php]: So, a fragment is like a template, and these are like variables? How about `fragmentVariables`, or just `variables` ?

Comment (srascar) [inline Helper/FragmentHelper.php]: perfect

Comment (qsomazzi) [inline Helper/FragmentHelper.php]: as discuss with @srascar those props are more a fieldCollection than variables, because we have to configure them on the BO. So we agreed on naming this `fields` it makes moire sence

Comment (greg0ire) [inline Helper/FragmentHelper.php]: It can also viewed as the `fields` of a class (after all, it looks like you're defining a "class", with fields), so, this name is fine too IMO. Also, you often say that fragments are like blocks, but why didn't you reuse the block bundle exactly?

Comment (qsomazzi) [inline Helper/FragmentHelper.php]: because the goal was to provide really simple fragments, a lot simpler than blocks. Also block are used to structure the page, fragments are just a collection we need to iterate over and display.

Comment (greg0ire) [inline Helper/FragmentHelper.php]: Thanks for the explanation

Comment (srascar) [inline Helper/FragmentHelper.php]: done

**Comment by qsomazzi** [comment]
seems all good for me 👍 @sonata-project/contributors ?

File: Helper/FragmentHelper.php [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+ */
+
+namespace Sonata\ArticleBundle\Helper;
+
+use Sonata\ArticleBundle\FragmentService\FragmentServiceInterface;
+use Sonata\ArticleBundle\Model\FragmentInterface;
+use Symfony\Component\Templating\EngineInterface;
+
+/**
+ * Class FragmentHelper.
```
Comment (greg0ire) [inline Helper/FragmentHelper.php]: Duh!.
Please improve or remove.

File: Resources/doc/reference/custom_fragment.rst [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+         * {@inheritdoc}
+         */
+        public function getTemplate()
+        {
+            return 'AcmeDummyBundle:Fragment:fragment_my_awesome.html.twig';
+        }
+    }
+
+
+Using ``settings..keys`` in ``buildEditForm`` method, you can define all elements that compose your fragment.
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: What's with the `..` notation? What do you mean exactly with it?

File: Resources/doc/reference/custom_fragment.rst [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+Every key is a field that is displayed on fragment edit form.
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: "the fragment edit form"

File: Resources/doc/reference/custom_fragment.rst [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+The ``getTemplate`` method allows you to define which template should be used when rendering this fragment type.
+
+
+Then you need to declare the service
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: "Then, you need to declare the service:"

File: Resources/doc/reference/custom_fragment.rst [hunk 25]
```diff
@@ -0,0 +1,90 @@
+Create a Custom Fragment
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: "Creating"

File: Resources/doc/reference/custom_fragment.rst [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+========================
+
+
+Define a fragment Service
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: "Defining"

File: Resources/doc/reference/custom_fragment.rst [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+        acme.dummy.fragment.my_awsome:
+            class: Acme\DummyBundle\FragmentService\MyAwesomeFragmentService
+            arguments:
+                - My Awesome Fragment Name # Fragment name in admin interface
+            tags:
+                - { name: sonata.article.fragment, key: acme.dummy.fragment.my_awesome }
+                # Where key is your fragment type unique identifier
+
+
+Define a fragment template
```
Comment (greg0ire) [inline Resources/doc/reference/custom_fragment.rst]: "Defining"

File: Helper/FragmentHelper.php (hunk 21 above)
Comment (srascar) [inline Helper/FragmentHelper.php]: This is inspired from SonataBlock `Sonata\BlockBundle\Templating\Helper\BlockHelper`

Comment (greg0ire) [inline Helper/FragmentHelper.php]: ... which I did not review. Do try to copy existing style, but don't assume existing code is perfect. This comment adds zero information, it's just noise,  so please improve or just remove it.

Comment (srascar) [inline Helper/FragmentHelper.php]: I thought you were talking about the class name sorry

Comment (greg0ire) [inline Helper/FragmentHelper.php]: Oh… I should have been more explicit from the start, my bad.

**Comment by srascar** [comment]
@greg0ire  i did closed the PR by error
please see https://github.com/sonata-project/SonataArticleBundle/pull/16

**Comment by greg0ire** [comment]
Not sure why github prevents you from reopening this one : apparently it wants new commits…
